coordinator/src/orderbook/mod.rs

```diff
@@ -6,3 +29,253 @@ pub mod websocket;
 
 #[cfg(test)]
 mod tests;
+
+pub struct Orderbook {
+    pool: Pool<ConnectionManager<PgConnection>>,
+    shorts: Shorts,
+    longs: Longs,
+}
+
+pub struct Shorts(Vec<Order>);
+
+impl Shorts {
+    pub fn add(&mut self, order: Order) {
+        self.0.push(order);
+    }
+    pub fn remove(&mut self, order_id: Uuid) {
+        self.0.retain(|o| o.id != order_id)
+    }
+    pub fn sort(&mut self) {
+        self.0.sort_by(|a, b| {
+            if a.price.cmp(&b.price) == Ordering::Equal {
+                return a.timestamp.cmp(&b.timestamp);
+            }
+            // Descending order.
+            b.price.cmp(&a.price)
+        })
+    }
+}
+
+pub struct Longs(Vec<Order>);
+
+impl Longs {
+    pub fn add(&mut self, order: Order) {
+        self.0.push(order);
+        self.sort();
+    }
+    pub fn remove(&mut self, order_id: Uuid) {
+        self.0.retain(|o| o.id != order_id)
+    }
+    pub fn sort(&mut self) {
+        self.0.sort_by(|a, b| {
+            if a.price.cmp(&b.price) == Ordering::Equal {
+                return a.timestamp.cmp(&b.timestamp);
+            }
+            // Ascending order.
+            a.price.cmp(&b.price)
+        });
+    }
+}
+
+impl Orderbook {
+    pub fn new(pool: Pool<ConnectionManager<PgConnection>>) -> Self {
+        Self {
+            pool,
+            shorts: Shorts(vec![]),
+            longs: Longs(vec![]),
+        }
+    }
+
+    async fn initialize(&mut self) -> Result<()> {
+        let all_orders = spawn_blocking({
+            let mut conn = self.pool.clone().get()?;
+            move || {
+                let all_orders =
+                    orders::get_all_orders(&mut conn, OrderType::Limit, OrderState::Open, true)?;
+                anyhow::Ok(all_orders)
+            }
+        })
+        .await??;
+
+        for order in all_orders {
+            match order.direction {
+                commons::Direction::Short => self.shorts.add(order.clone()),
+                commons::Direction::Long => self.longs.add(order.clone()),
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn add_order(
+        &mut self,
+        new_order: NewOrder,
+        order_reason: OrderReason,
+    ) -> Result<Option<Message>> {
+        let order = spawn_blocking({
+            let mut conn = self.pool.clone().get()?;
+            move || {
+                let order = match new_order {
+                    NewOrder::Market(o) => {
+                        orders::insert_market_order(&mut conn, o.clone(), order_reason)
+                    }
+                    NewOrder::Limit(o) => orders::insert_limit_order(&mut conn, o, order_reason),
+                }
+                .map_err(|e| anyhow!(e))
+                .context("Failed to insert new order into DB")?;
+
+                anyhow::Ok(order)
+            }
+        })
+        .await??;
```

Comment:
I am afraid if we simply do not await we might miss errors or even inserts e.g. when the coordinator restarts before the insert is done.

I was thinking of a task running e.g. every few seconds could sync the in memory orderbook with the database. On a graceful shutdown we can then sync any pending updates to ensure we do not miss any orders.

Reply:
I think we need to prioritise speed over everything else. Restarts are very rare and we could handle them gracefully.

In case of a panic, I am not even sure we can easily run into trouble.